contributors/design-proposals/federation-clusterselector.md

```diff
@@ -0,0 +1,78 @@
+
+In situations where objects are not to be forwarded to federated clusters, instead a delete api call will be made using the object definition. If the object does not exist it will be ignored.
+
+The federation-controller will be used to implement this with shared logic stored as utility functions to reduce duplicated code where appropriate.
+
+### End User Functionality 
+The annotation `federation.alpha.kubernetes.io/cluster-selector` is used on kubernetes objects to specify additional placement decisions that should be made. The value of the annotation will be a json object of type ClusterSelector which is an array of type ClusterSelectorRequirement.
+
+Each ClusterSelectorRequirement is defined in three possible parts consisting of
+1. Key - Matches against label keys on the Federated Clusters.
+1. Operator - Represents how the Key and/or Values will be matched against the label keys and values on the Federated Clusters one of ("In", in", "=", "==", "NotIn", notin", "Exists", "exists", "!=", "DoesNotExist", "!", "Gt", "gt", "Lt", "lt").
+1. Values - Matches against the label values on the Federated Clusters using the Key specified. When the operator is "Exists", "exists", "DoesNotExist" or "!" then Values should not be specified.
+
+Example ConfigMap that uses the ClusterSelector annotation. The yaml format is used here to show that the value of the annotation will still be json.
+```yaml
+apiVersion: v1
+data:
+  myconfigkey: myconfigdata
+kind: ConfigMap
+metadata:
+  annotations:
+    federation.alpha.kubernetes.io/cluster-selector: '[{"key": "location", "operator":
+      "in", "values": ["europe"]}, {"key": "environment", "operator": "==", "values":
+      ["prod"]}]'
+  creationTimestamp: 2017-02-07T19:43:40Z
+  name: myconfig
+```
+
+In order for the configmap in the example above to be forwarded to any Federated Clusters they MUST have two Labels: "location" with at least one value of "europe" and "environment" that has a value of "prod".
+
+### Matching Logic
+
+The logic to determine if an object is sent to a Federated Cluster will have two rules.
+
+1. An object with no `federation.alpha.kubernetes.io/cluster-selector` annotation will always be forwarded on to all Federated Clusters even if they have labels configured. (this ensures no regression from existing functionality)
+
+1. If an object contains the `federation.alpha.kubernetes.io/cluster-selector` annotation then ALL ClusterSelectorRequirements must match in order for the object to be forwarded to the Federated Cluster.
+
```

Comment:
ok, this isn't handled at all in the current pull request so if replicaset preferences are applied before it reaches each individual controller then it currently sets it first, then possibly gets filtered by the cluster selector. In either case there would possibly be a mismatch between the number of replica sets asked for and the number actually created so we should perhaps recommend that folks use one method or the other and not both.

let me know your thoughts @mwielgus .